src/rpc/auth_api.rs

```diff
@@ -1,18 +1,24 @@
 // Copyright 2019-2023 ChainSafe Systems
 // SPDX-License-Identifier: Apache-2.0, MIT
+#![allow(clippy::redundant_allocation)]
 
 use crate::auth::*;
 use crate::lotus_json::LotusJson;
+use crate::rpc::error::JsonRpcError;
 use crate::rpc_api::{auth_api::*, data_types::RPCState};
+use anyhow::Result;
 use fvm_ipld_blockstore::Blockstore;
-use jsonrpc_v2::{Data, Error as JsonRpcError, Params};
+use jsonrpsee::types::Params;
+
+use std::sync::Arc;
 
 /// RPC call to create a new JWT Token
 pub async fn auth_new<DB: Blockstore>(
-    data: Data<RPCState<DB>>,
-    Params(params): Params<AuthNewParams>,
+    params: Params<'_>,
+    data: Arc<Arc<RPCState<DB>>>,
```

Comment:
The first `Arc` is needed to share the state between two RPC modules.
https://github.com/ChainSafe/forest/blob/ce33815f8ffb9670a4f31bb327027b2ebbed3f3d/src/rpc/mod.rs#L69

We only have one module now, but to implement specific pubsub code, we will need to implement a dedicated `RpcModule`.
In the context of `async` methods, `jsonrpsee` creates a second `Arc` around the state.

Reply:
Yes, because all methods in a module need access to the state.

In our case, we will have to call `chain_api::chain_notify` to get the subscriber:

```
async fn chain_notify<DB: Blockstore>(
     data: Arc<RPCState<DB>>,
 ) -> Result<Subscriber<HeadChange>, JsonRpseeError> {
     let head_change = data.chain_store.publisher().subscribe();

     Ok(head_change)
 }
```

Reply:
Hum, there's only one `Arc` here.

Reply:
Ah I see why, ignore my comment.

Reply:
What's the resolution here? this is the only thing blocking an approval from me I think

Reply:
I think the double `Arc<Arc<T>>` is ok for now (I've hidden the ugliness using the `Data` type alias). This should resolve at the `jsonrpee` crate level. Maybe with something like:

```
pub fn from_arc(ctx: Arc<Context>) -> Self {
  Self { ctx, methods: Default::default() }
}
```

Reply:
I'm grumpy about this, but if there's truly no other way then fine

Reply:
The `from_arc` works and I can remove one Arc level: `pub type Data<T> = Arc<T>;`

This will require change in upstream repository though.